### PR TITLE
feat(swarm): introduce `ConnectionId::new_unchecked` constructor

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

In earlier iterations of the design for generic connection management, we removed the `ConnectionId::new` constructor because it would have allowed users to create `ConnectionId`s that are already taken, thus breaking invariants that `NetworkBehaviour`s rely on. Later, we incorporated the creation of `ConnectionId` in `DialOpts` which mitigates this risk altogether.

Thus, it is reasonably safe to introduce a public, non-deprecated constructor for `ConnectionId` that can be used for tests.

Related https://github.com/libp2p/rust-libp2p/pull/3327#issuecomment-1469870307.

## Notes & open questions

cc @melekes 

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/CHANGELOG.md
Comment (melekes) [inline swarm/CHANGELOG.md]: PR number and link are missing

**Review by melekes — APPROVED** [review]

**Comment by thomaseizinger** [comment]
I am going to directly release this after it merges to unblock @melekes. This is trivial in my eyes, hence I am merging this right away.

File: swarm/src/connection.rs
Comment (mxinden) [inline swarm/src/connection.rs:76]:
``` diff
-     pub fn new_unchecked(id: usize) -> Self {
+     pub unsafe fn new_unchecked(id: usize) -> Self {
```

Wouldn't this be a good place for `unsafe`, similar how it is used in `NonZeroUsize`?

https://doc.rust-lang.org/stable/std/num/struct.NonZeroUsize.html#method.new_unchecked

Comment (thomaseizinger) [reply swarm/src/connection.rs:76]: > ```diff
> -     pub fn new_unchecked(id: usize) -> Self {
> +     pub unsafe fn new_unchecked(id: usize) -> Self {
> ```
> 
> Wouldn't this be a good place for `unsafe`, similar how it is used in `NonZeroUsize`?
> 
> [doc.rust-lang.org/stable/std/num/struct.NonZeroUsize.html#method.new_unchecked](https://doc.rust-lang.org/stable/std/num/struct.NonZeroUsize.html#method.new_unchecked)

I don't think so. `unsafe` is for memory-unsafety. AFAIK, `NonZeroUsize` relies on memory optimizations that require the value to never be zero, thus there is actually UB if you initialize it with `0`.

In our case, no UB occurs, you may just run into buggy `NetworkBehaviour`s.

Comment (mxinden) [reply swarm/src/connection.rs:76]: Document out-of-band conversation. I missed that a user can never inject a custom `ConnectionId` into `Swarm`.

**Comment by melekes** [comment]
@thomaseizinger @mxinden what is the timeline for the next patch release (0.51.2)? we're (Parity) interested in this change specifically. I can create a PR if necessary.

**Comment by mxinden** [comment]
I will try tomorrow. No promises @melekes.

**Pull request merged**
